src/java/htsjdk/samtools/util/AsyncBufferedIterator.java

```diff
@@ -0,0 +1,218 @@
+	 * 
+	 * @param iterator
+	 *            iterator to traverse
+	 * @param threadName
+	 *            background thread name
+	 * @param bufferCount
+	 *            number of read-ahead buffers
+	 * @param batchSize
+	 *            size of each read-ahead buffer. A larger batch size will
+	 *            increase both throughput and latency.
+	 */
+	public AsyncBufferedIterator(Iterator<T> iterator, String threadName, int bufferCount, int batchSize) {
+		if (iterator == null)
+			throw new IllegalArgumentException();
+		if (bufferCount <= 0 || batchSize <= 0)
+			throw new IllegalArgumentException("Buffer size must be at least 1.");
+		this.underlying = iterator;
+		this.buffer = new ArrayBlockingQueue<List<T>>(bufferCount);
+		this.batchSize = batchSize;
+		this.readerRunnable = new ReaderRunnable();
+		this.reader = new Thread(readerRunnable,
+				threadName == null ? getThreadNamePrefix() + threadsCreated.incrementAndGet() : threadName);
+		this.reader.setDaemon(true);
+		log.debug("Starting thread " + this.reader.getName());
+		this.reader.start();
+	}
+
+	protected String getThreadNamePrefix() {
+		return "AsyncBufferedIterator";
+	}
+
+	@Override
+	public void close() {
+		closeCalled = true;
+		try {
+			reader.interrupt();
+			buffer.clear(); // flush buffer so EOS indicator can be written if
+							// writer is blocking
+			reader.join();
+		} catch (InterruptedException ie) {
+			throw new RuntimeException("Interrupted waiting for background thread to complete", ie);
+		} finally {
+			CloserUtil.close(underlying);
+		}
+	}
+
+	@Override
+	public boolean hasNext() {
+		throwOnCallingThread();
+		if (closeCalled)
+			return false;
+		if (currentList == null || (currentList.size() > 0 && !currentBuffer.hasNext())) {
+			try {
+				currentList = buffer.take();
+				currentBuffer = currentList.iterator();
+			} catch (InterruptedException e) {
+				throw new RuntimeException("Error reading from background thread", e);
+			}
+			// rethrow any exceptions raised on the background thread while we
+			// were blocking on the next record
+			throwOnCallingThread();
+		}
+		return currentBuffer.hasNext();
+	}
+
+	@Override
+	public T next() {
+		if (hasNext())
+			return (T) currentBuffer.next();
+		throw new NoSuchElementException("next");
+	}
+
+	/**
+	 * Rethrows any exception encountered on the background
+	 * thread to the caller 
+	 */
+	private final void throwOnCallingThread() {
+		final Throwable t = this.ex.get();
+		if (t != null) {
+			if (t instanceof Error) {
+				throw (Error) t;
+			} else if (t instanceof RuntimeException) {
+				throw (RuntimeException) t;
+			} else {
+				throw new RuntimeException(t);
+			}
+		}
+	}
+
+	/**
+	 * Reads the given iterator, passing back records to the calling thread in batches
+	 */
+	private class ReaderRunnable implements Runnable {
+		public void run() {
+			boolean eosWritten = false;
+			try {
+				while (underlying.hasNext()) {
+					List<T> readAhead = new ArrayList<T>(batchSize + 1);
```

Comment:
my last `final` comment, but loading this into an IDE like `IntelliJ` should give you markers where you can add `final` and such.